tsdb/chunks/chunks.go

```diff
@@ -284,27 +301,85 @@ func MergeChunks(a, b chunkenc.Chunk) (*chunkenc.XORChunk, error) {
 	return newChunk, nil
 }
 
+// WriteChunks writes as many chunks as possible to the current segment,
+// cuts a new segment when the current segment is full and
+// writes the rest of the chunks in the new segment.
 func (w *Writer) WriteChunks(chks ...Meta) error {
-	// Calculate maximum space we need and cut a new segment in case
-	// we don't fit into the current one.
-	maxLen := int64(binary.MaxVarintLen32) // The number of chunks.
-	for _, c := range chks {
-		maxLen += binary.MaxVarintLen32 + 1 // The number of bytes in the chunk and its encoding.
-		maxLen += int64(len(c.Chunk.Bytes()))
-		maxLen += 4 // The 4 bytes of crc32
+	var (
+		batchSize  = int64(0)
+		batchStart = 0
+		batches    = make([][]Meta, 1)
+		batchID    = 0
+		firstBatch = true
+	)
+
+	for i, chk := range chks {
+		// Each chunk contains: data length + encoding + the data itself + crc32
+		batchSize += int64(MaxChunkLengthFieldSize) // The data length is a variable length field so use the maximum possible value.
+		batchSize += ChunkEncodingSize              // The chunk encoding.
+		batchSize += int64(len(chk.Chunk.Bytes()))  // The data itself.
+		batchSize += crc32.Size                     // The 4 bytes of crc32.
+
+		// Cut a new batch when it is not the first chunk(to avoid empty segments) and
+		// the batch is too large to fit in the current segment.
+		cutNewBatch := (i != 0) && (batchSize+SegmentHeaderSize > w.segmentSize)
+
+		// When the segment already has some data the
+		// the first batch size calculation should account for that.
+		if firstBatch && w.n > SegmentHeaderSize {
+			cutNewBatch = batchSize+w.n > w.segmentSize
+			if cutNewBatch {
+				firstBatch = false
+			}
+		}
+
+		if cutNewBatch {
+			batchStart = i
+			batches = append(batches, []Meta{})
+			batchID++
```

Comment:
reset `batchSize` ?

Reply:
Add more chunks to test case 1:
```
		// 1:Two chunks can fit in a single segment can fit so the last one should result in a new segment.
		{
			chks: [][]chunks.Meta{
				[]chunks.Meta{
					chk1,
					chk2,
					chk3,
					chk4,
					chk5,
				},
			},
		// . . .
		},
```
batches: 
```
[[{Ref:0 Chunk:0xc00015a300 MinTime:1 MaxTime:1} {Ref:0 Chunk:0xc00015a320 MinTime:1 MaxTime:1}] [{Ref:0 Chunk:0xc00015a340 MinTime:1 MaxTime:1}] [{Ref:0 Chunk:0xc00015a360 MinTime:1 MaxTime:1}] [{Ref:0 Chunk:0xc00015a380 MinTime:1 MaxTime:1}]] 
```

Reply:
:) and when I thought I fixed all cases.
Updated.

Reply:
Nice!